### PR TITLE
Swap

### PR DESCRIPTION
**Pull request by egilkv**
Fix bug in Serial.swap() that made it not work at all.

Additionally, implemented swap() so that it can invoked before begin(), so there is no bad effects on the original tx/rx pins.

### CODE REVIEW

**Comment by MCUdude** [comment]
I prefer a separate Serial.swap function instead of the swap functionality being a third parameter to Serial.begin(). They do it like this with ESP8266.

It feels kinda backwards to have to add 8N1 as a second parameter in order for the third one to work. Can't this be solved while still keeping Serial.swap()?

**Comment by egilkv** [comment]
I do understand your griefs with a third parameter to Serial.begin().

(As a digression I guess it is a design bug in C++ that you cannot do 
stuff like Serial.begin(9600,,true); )

There are ESP32 variants that have used additional parameters for 
Serial.begin() for other purposes, and it would create extra confusion 
if it is different here. (ESP32 has no swap, btw)

Regardless, the swap() function, as implemented here and for the 
ESP8266, is simply a really ugly hack and a bad design. It redefines as 
INPUTs pins it has no business touching, which can lead to bugs the user 
would not be aware of. On top of all of this, it does not even work as a 
dynamic pin swap function: 
https://github.com/esp8266/Arduino/issues/2380 This is due to two 
things, firstly that it does not sync with transmission/reception but 
also that it leaves pins in tri-state when they should not. (The proper 
way to do dynamic switching between two set of pins like this is to do 
an end() first, then change pin assignments, then do a new begin().)

The absolutely best way to assign pins would be in the constructor, like 
for the SoftwareSerial library. However, in the effort to make "hello 
world" as simple as possible, the constructor for the Serial is hidden.

So lets go for the next best solution. I found the ESP32 also has a 
method called pins() that defines which pins to use. That to me sounds 
like the best way forwards:

     1. It will work nicely with possible future chips that may have 
even more flexible options for pin swopping.

     2. It follows a pattern familiar with other Arduino classes.

     3. It is intuitive to use

An example:

#define PIN_TXD1 18
#define PIN_RXD1 19

   Serial1.pins(PIN_TXD1,PIN_RXD1);
   Serial1.begin(4800);

The #defines are of course optional, but I like to make a list of all my 
pin usage in one place.

I plan to implement this also for Wire and SPI.

Egil

**Comment by egilkv** [comment]
I have now implemented pins() for Wire and for SPI, and also added a section in the README.

**Comment by MCUdude** [comment]
Sorry, I'm not trying to be difficult, but pins() with two parameters is IMO just overly complicated for what it is. None of the megaAVR-0 microcontrollers supports arbitrary pin swapping like the pins() function indicates. IMO it's not necessary to specify the actual pins at all, but it would be better if we indicated what mode to switch to instead.

Here's an example of what I'd prefer:

``` c++

#define DEFAULT_PINS        0
#define ALTERNATIVE_PINS    1
#define ALTERNATIVE_PINS_1  1
#define ALTERNATIVE_PINS_2  2

Serial.swap();                   // No parameter -> always swap to alternative pins
Serial.swap(ALTERNATIVE_PINS);   // Swap to alternative pins
Serial.swap(ALTERNATIVE_PINS_1); // Swap to alternative pins
Serial.swap(DEFAULT_PINS);       // Swap to default pins

Wire.swap();                   // No parameter -> always swap to alternative pins
Wire.swap(ALTERNATIVE_PINS);   // Swap to alternative pins
Wire.swap(ALTERNATIVE_PINS_1); // Swap to alternative pins
Wire.swap(DEFAULT_PINS);       // Swap to default pins

SPI.swap();                   // Will not work because we have more that one option to swap to
SPI.swap(ALTERNATIVE_PINS);   // Swap to alternative pins 1
SPI.swap(ALTERNATIVE_PINS_1); // Swap to alternative pins 1
SPI.swap(ALTERNATIVE_PINS_2); // Swap to alternative pins 2
SPI.swap(DEFAULT_PINS);       // Swap to default pins
```

**Comment by egilkv** [comment]
Did you read what I wrote about how swap() is fundamentally flawed? Also in the original ESP8266 implementation. For instance:

pinMode(0, OUTPUT)
Serial.begin(9600);
Serial.swap(ALTERNATIVE_PINS):

will leave pin 0 as an INPUT.

I have implemented and tested the pins() for the relevant devices, and I think you'll find the code is quite compact. The choice is yours, of course, but I will not touch the swap().

**Comment by MCUdude** [comment]
I understand that the current swap implementation is flawed and have to be improved/replaced.
but do we actually need to specify exactly what pins we are swapping to?

Couldn't Serial.swap look like this instead (while keeping the rest of your code)?

``` c++
bool UartClass::swap(uint8_t muxSetting)
{
  if(muxSetting == DEFAULT_PINS)
  {
    _pin_set = 0;
    return true;
  }
  else if(muxSetting == ALTERNATIVE_PINS_1) // or ALTERNATIVE_PINS
  {
    _pin_set = 1;
    return true;
  }
  else
  {
    _pin_set = 0; // Default to standard
    return false;
  }
}
```

**Comment by egilkv** [comment]
I am sorry, but the alternative you suggest adds complexity, and introduces new terms like ALTERNATIVE_PINS_1 for the user. Which means they have to look up the data sheet to find out what that really means. This is contrary to Arduino philosophy. 
By specifying which pins to use, the user only has to relate to things he already knows, and does not have to refer to a data sheet. So it is easier. As I have shown, listing also a pattern that appears with many other Arduino classes.
Finally, I have now used this concept in a project that uses a m4809 with 4 serial ports, external 32kHz and external clock, I2C, SPI, and numerous analog, digital and PWM pins, pretty much using every pin. This is how I discovered all the issues in your code. The project heavily depends on pin swaps, and believe me, actually listing the pins is the best way of keeping track of all these pins. 

I do not want to spend more time on this issue. If you believe your insight into software design is better, by all means go ahead. I have implemented the pins() feature that works nicely, and I will definitely not turn back.

**Comment by MCUdude** [comment]
I think we can agree that finding what pin swaps go to where in the datasheet isn't particularly easy for the average user. That's why I've made a simplified pinout diagram where alternative swaps are clearly shown.

> I am sorry, but the alternative you suggest adds complexity and introduces new terms like ALTERNATIVE_PINS_1 for the user

On one side we already know what physical pins we're swapping to, so having them as two arguments isn't really a big problem. On the other side, the user has to make sure he/she gets the pin order right (tx, rx) based on the current pinout in use.

The benefit with swap is that all the user needs to care about is what swap position to select. You can easily find out what actual pins this corresponds to in the pinout diagram.
Also, we don't have to use macros as all. Serial.swap(0) and Serial.swap(1) will also do the job.
I think I'm heading towards including both, just like the ESP8266 Arduino core does. And swap will be based on your implementation.

You should try to open a PR on the main repo to see if you can get this merged there as well. It's a very useful feature indeed, and it's too bad it's not officially implemented in the new Arduino API.

File: README.md [hunk 1]
```diff
@@ -106,6 +107,53 @@ Please have a look at the pins_arduino.h files for detailed info.<br/> <br/>
...
 
+## Alternative pins
+The megaAVR-0 microcontrollers supports alternative pin assignments
+for some of the built in peripherals.
+This is specified by invoking the `pins()` method before the `begin()` for the associated peripheral.
+The `pins()` method will return `true` if the pin combination is supported.
+For `Serial` peripherals the method is `pins(tx,rx)`. 
+(Note that this is the same pin sequence as used for the
+ESP8266 `pins` method, but the opposite of the one SoftwareSerial uses.)
+For `Wire` the method is `pins(sda,scl)`, and for `SPI` it is `pins(mosi,miso,scl,ss)`.
```
Comment (MCUdude) [inline README.md]: The clock line on SPI is commonly called sck instead of scl

**Comment by egilkv** [comment]
Mea culpa, now corrected. Thanks

**Comment by MCUdude** [comment]
I'm going through the code in the PR to test it properly. I get this warning whenever I try to compile a sketch where Wire.h is included:

```
from /Users/hans/Documents/Arduino/hardware/MegaCoreX/megaavr/cores/coreX-corefiles/UART3.cpp:25:
/Users/hans/Documents/Arduino/hardware/MegaCoreX/megaavr/variants/28pin-standard/pins_arduino.h:76:16: warning: 'SCL' defined but not used [-Wunused-variable]
 static uint8_t SCL = PIN_WIRE_SCL;
                ^~~
/Users/hans/Documents/Arduino/hardware/MegaCoreX/megaavr/variants/28pin-standard/pins_arduino.h:75:16: warning: 'SDA' defined but not used [-Wunused-variable]
 static uint8_t SDA = PIN_WIRE_SDA;
```

Is it OK if add the `unused` attribute to SDA and SCL? Are there more elegant ways to do this? Adding `(void)SDA;` and `(void)SCL;` to the TwoWire::begin() methods didn't have any effect.

```c
__attribute__ ((unused)) static uint8_t SDA = PIN_WIRE_SDA;
__attribute__ ((unused)) static uint8_t SCL = PIN_WIRE_SCL;
```

**Comment by MCUdude** [comment]
Turns out that `Wire.pins()` doesn't change SDA and SCL values at all. Any idea why?

**Pull request closed without merging**